kafka_helpers/consumer.go

```diff
@@ -0,0 +1,65 @@
+package kafka_helpers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/segmentio/kafka-go"
+	"internship_project/elasticsearch_helpers"
+	"log"
+)
+
+type KafkaConsumer struct {
+	Reader   *kafka.Reader
+	EsClient elasticsearch_helpers.ElasticsearchClient
+}
+
+func (consumer *KafkaConsumer) Consume() {
+	fmt.Println("KafkaConsumer is ready to consume new messages.")
+	for {
+
+		m, err := consumer.Reader.FetchMessage(context.Background())
+		if err != nil {
+			log.Printf("Error while fetching message")
+		}
+
+		fmt.Printf("message at offset %d: %s = %s\n", m.Offset, string(m.Key), string(m.Value))
+
+		var jsonMessage map[string]interface{}
+		err = json.Unmarshal(m.Value, &jsonMessage)
+		if err != nil {
```

Comment:
Now that the message is no longer skipped, there is still an issue of the stream getting stuck. I'd like you to propose a solution for this issue.

Reply:
We've added another consumer that will consume messages from "retry" topic.
- Every time an error pops up, for whatever reason, the message will be commited on main topic, but not written to ElasticSearch, just so we can skip it and continue with consuming new messages
- Then, it will be written to the "retry" topic, which will be consumed every 15s (time period is intentionally super short)
- If the retry consumer fails as well, we will skip the message altogether

Reply:
It's better if you extract time period into a config file. You'll have better control of the retry mechanism. There can be situations where Kafka, Zookeper and/or Elasticsearch can be unreachable ( crashed, index stuck, connection lost, etc ). Was this taken into account? Or was the focus on retry error handling caused by some invalid data?

Reply:
I want to add that the "retry" topic is a great idea and how it is usually handled. It's more commonly known as deadletter queue/topic. You just don't have to consume it regularly and can just leave it as is, until you have resolved the issue and can manually run the recovery process. Just don't skip messages in the deadletter topic.

Reply:
@Pajovic 
- That does make a lot of sense, next time we will hide it within a config file, it should serve us well as it is right now.
- We did not take those issues into account, since we were not able to reproduce them manually and work on them, therefore we just protected the consumers from any errors that might occur on all operations
- Invalid data coming to the consumer won't be an issue since data only reaches Kafka when it's successfully manipulated within our app.

@lazar-nikolic-ava so you propose that we shouldn't consume them at all? How are we supposed to handle them in that case? Our idea was to consume them on a regular period (e.g. 1 hour, it was set to 15s here for simplicity), so that the data does get to the main topic at some point. Maybe we shouldn't have skipped the message in retry if we failed to consume it from there, a better solution would to permanently save it to another deadletter topic for a longer period of time, without consuming. How does that sound to you?

Reply:
@igorroncevic You can test that by starting your service, and when everything is initialised, shutdown Kafka and try to make a request. These scenarios are very real in any serious project and need to be handled to avoid data loss. So it's better if you're aware of them further on.

Reply:
@igorroncevic Regarding the dead letter topic, I guess it all depends on the error type. Maybe it would be a good idea to check why fetch or commit or any action towards elastic search failed before publishing data to dead letter. You can avoid unnecessary publishes to dead letter this way. Connection lost comes first to my mind. If however something is meant to be published to dead letter, you can check the nature of the problem and, when fixed, try to consume from that topic manually or periodically, but don't discard anything from it until processed successfully.

Reply:
I extracted all Kafka and Elasticsearch parameters in conf file. We also tested what happens when Kafka is down and request is made - error is logged but our service does not exit. We have data loss there but we decided to cover only scenario where error is caused by invalid data. We hope that is ok for out project.

Reply:
@igorroncevic My opinion is to leave erroneous messages in the deadletter topic. Once you have a fix for them, run a script or hit an endpoint that moves messages from the deadletter topic onto the regular topic.

Reply:
As Pajović said, don't discard anything from the deadletter topic. Remove them only once you have run the recovery procedure and messages were successfully processed, i.e. committed to the regular topic.

Reply:
@zukovicfilip Errors caused by invalid data is exactly what we're trying to fix with the deadletter topic. If you're committing stuff to the deadletter topic, you don't need to do anything else for this project. We were just discussing the mechanisms of recovery using a deadletter topic.

Reply:
I've added an endpoint that a logged user can hit and transfer messages from the deadletter queue to the main topic.
Key things I'd like point out:
- I was not able to get the amount of messages that is supposed to be read, therefore I couldn't know when to stop consuming. Offsets get messed up a bit when they need to be read from a consumer group. For that, I allowed the consumer to fail at working with the topic for a limited number of times and disconnecting afterwards
- I wanted to ask whether the approach that I tried at first with `context.WithTimeout` was good enough. With that, I wanted to check whether the consumer is taking too long to read from the topic, dismissing his attempt if it took more than time defined in contextDeadline. It was kind of unnecessary after I added the counters from the previous point, but it seemed like a pretty good solution at first. The only issue with that one is perhaps the connection is really taking too long and the context might not wait enough for it to succeed, therefore failing too soon.

Reply:
I like that context approach. You can set a configurable timeout with it and pass it as a ctx param all the way to kafka driver. This way you have control on how long you expect to, at least, try consuming data, and all resources are handled and/or released properly - you don't have to count anything, the driver does that for you. That's one of the best things ctx provides and that's why it's used.

Reply:
Have you checked how I've done it? I assumed it would be much better than counting, but I wasn't sure it was good enough due to the reason I mentioned above. Is that what you meant? I'm basically sending that configured context to the FetchMessage, which properly quits trying after a certain amount of time and that is exactly what we need.

Reply:
From what I'm seeing, it looks like deadletter recovery will attempt to fetch messages for a set number of times even when the deadline is exceeded. I came to the conclusion by reading the code, not testing it so I could pretty much be wrong. I wouldn't worry too much about it now anyway. For future reference, it might be a good idea to just stop if the deadline has been reached. Looks pretty good for our little project here.

Reply:
This has been fixed in the latest commit and now only the deadline is being watched.